METRON-1382 Run Stellar in a Zeppelin Notebook

nickwallen opened the pull request:
## Stellar in a Zeppelin Notebook

The goal of this effort was to have the shell-like experience of the command-line driven REPL, but inside a web-based Zeppelin Notebook.  This makes the existing functionality provided by the REPL more accessible and approachable for the average user.  

This also opens up interesting additional scenarios like "executable" use cases or plotting profile data.  Definitely more work to do on those fronts, but this is a first step.

### Changes

1. There was a good amount of refactoring that I needed to do to achieve this.  Prior to this effort different aspects of the REPL's functionality were tied together in the `StellarShell` class.  To be able to have the same shell-like experience in both the REPL and Zeppelin, I needed to factor this logic out separately to allow for reuse.  

    This included the following bits.
     * Console logic tied to the AESH library
     * Stellar execution logic
     * Autocomplete logic
     * Variable assignment (only available in the shell)
     * Magics
     * Docstrings
     * Comments
     * `quit`

1. We had very few automated tests around the shell/REPL, primarily because all these bits were tied together, which made testing difficult.  Separating out this logic allowed me to add a large number of unit tests around each of these pieces of logic.



1. This also adds a separate project `metron-stellar/stellar-zeppelin` that provides a Zeppelin interpreter that allow you to run Stellar in a notebook.  The interpreter acts as the front-end that then re-uses all of the bits that I refactored as described above.

### What does this do?

This allows you to run Stellar expressions in a Zeppelin Notebook.

This provides auto-complete in the Zeppelin Notebook using "CTRL + ."

### What does this not do?

This does not automatically prepare the Zeppelin environment to run Stellar.  Meaning it does not install the Stellar interpreter in Zeppelin.  You have to manually install the zinterpreter in Zeppelin by following the directions described in the `metron-stellar/stellar-zepelin` README.

### Testing

Run Stellar in Notebook
1. Build Metron.
1. Follow the instructions in `metron-stellar/stellar-zepelin/README.md` to download/run Zeppelin locally.
1. Follow the same instructions to execute Stellar expressions in a Zeppelin Notebook.
1. Use CTRL+. to try out autocomplete.
    
Validate the CLI REPL
1. Deploy Full Dev
1. Run the existing CLI-driven REPL
1. Validate that we have feature parity


## Pull Request Checklist

- [ ] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [ ] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?
- [ ] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [ ] Have you included steps or a guide to how the change may be verified and tested manually?
- [ ] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
- [ ] Have you written or updated unit tests and or integration tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/DefaultStellarAutoCompleter.java:
```diff
@@ -0,0 +1,235 @@
+/**
+ * Provides auto-completion for Stellar.
+ */
+public class DefaultStellarAutoCompleter implements StellarAutoCompleter {
```
This class provides the core auto-completion logic for both the REPL and Zeppelin.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/DefaultStellarAutoCompleter.java:
```diff
@@ -0,0 +1,235 @@
+    this.autocompleteIndex = initializeIndex();
+
+    // TODO is this needed?  FunctionResolver functionResolver
+//    // asynchronously update the index with function names found from a classpath scan.
```
Is this logic needed any longer?  I don't feel it is, but need to dig on this some more.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/DefaultStellarShellExecutor.java:
```diff
@@ -0,0 +1,398 @@
+ * Default implementation of a StellarShellExecutor.
+ */
+public class DefaultStellarShellExecutor implements StellarShellExecutor {
+
```
This class provides the core execution logic for both the REPL and Zeppelin.  It handles core Stellar in addition to the "extensions" like assignment, magics, docstrings, etc.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/DefaultStellarShellExecutor.java:
```diff
@@ -0,0 +1,398 @@
+    for(SpecialCommand command : commandRegistry) {
+      if(command.getMatcher().apply(expression)) {
+        return command.execute(expression, this);
+      }
```
Anything that is not "core" Stellar is defined separately as a `SpecialCommand`.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/DefaultStellarShellExecutor.java:
```diff
@@ -0,0 +1,398 @@
+            new MagicDefineGlobal(),
+            new MagicUndefineGlobal(),
+            new MagicListGlobals()
+    );
```
All the specials are made available here.

nickwallen replied:
If we ever do something like add assignment to the core Stellar language (like #687) then all we have to do is not use/remove the `AssignmentCommand` class.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/StellarExecutionListeners.java:
```diff
@@ -0,0 +1,42 @@
+
+  public interface SpecialDefinedListener {
+    void whenSpecialDefined(SpecialCommand magic);
+  }
```
Event listeners can be notified when variables, functions, or specials are defined.  Useful in our case for notifying the autocompleter.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/cli/StellarShell.java:
```diff
@@ -0,0 +1,409 @@
+ *
+ * Useful for debugging Stellar expressions.
+ */
+public class StellarShell extends AeshConsoleCallback implements Completion {
```
The StellarShell now only contains logic to drive the Aesh console.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/cli/StellarShell.java:
```diff
@@ -0,0 +1,409 @@
+   */
+  @Override
+  public void complete(CompleteOperation completeOperation) {
+    String buffer = completeOperation.getBuffer();
```
Ties our generic autocompleter to the autocomplete interface required by Aesh.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/specials/SpecialCommand.java:
```diff
@@ -0,0 +1,60 @@
+   * @param executor A stellar execution environment.
+   * @return The result of executing the magic command.
+   */
+  StellarShellResult execute(String expression, StellarShellExecutor executor);
```
The interface for all specials; assignment, comments, magics, docstrings, etc.

nickwallen commented on metron-stellar/stellar-zeppelin/README.md:
```diff
@@ -0,0 +1,85 @@
+
+[Apache Zeppelin](https://zeppelin.apache.org/) is a web-based notebook that enables data-driven, interactive data analytics and collaborative documents with SQL, Scala and more.  This project provides a means to run the Stellar REPL directly within a Zeppelin Notebook.
+
+## Installation
```
Instructions for installing/running in a notebook.

justinleet replied:
Can you please add the license header to this? https://github.com/apache/metron/pull/884 is close to going in and enforcing this, so I'm hoping to avoid impact to master.

```
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->
```

nickwallen replied:
Sure, done.

nickwallen commented:
There are a few more things I want to clean-up on this, but its 95% there.

nickwallen commented on metron-platform/metron-management/src/main/java/org/apache/metron/management/ShellFunctions.java:
```diff
@@ -90,19 +95,19 @@ public Object apply(List<Object> args) {
         data[i++] = new String[] { toWrappedString(kv.getKey().toString(), wordWrap)
                                  , toWrappedString(result.getResult(), wordWrap)
-                                 , toWrappedString(result.getExpression(), wordWrap)
+                                 , toWrappedString(result.getExpression().get(), wordWrap)
```
The VariableResult.expression field is now optional.  We are not always know the expression that resulted in a value.  The ShellFunctions just needed updated to treat this as an Optional.

nickwallen commented on metron-stellar/stellar-common/README.md:
````diff
@@ -1377,7 +1377,7 @@ IS_EMAIL
 ```
 $ mvn exec:java \
-   -Dexec.mainClass="org.apache.metron.stellar.common.shell.StellarShell" \
+   -Dexec.mainClass="org.apache.metron.stellar.common.shell.cli.StellarShell" \
````
`StellarShell`, the main driver class for the CLI-based REPL, was moved to its own package since it is only used by the CLI-based REPL.   This separates it from the other core classes that are used by **both** the Zeppelin and CLI-based REPLs.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/StellarExecutionListeners.java:
```diff
@@ -0,0 +1,51 @@
+   */
+  public interface FunctionDefinedListener {
+    void whenFunctionDefined(StellarFunctionInfo functionInfo);
+  }
```
I used an event listener pattern so that external entities could get notified when things occur during the execution of Stellar.  For example, the auto-completer needs to know when a variable is defined.

ottobackwards replied:
Nothing seems to be ever removed, can that ever be a problem?  That will typically cause retention...  and 'leaks'.

nickwallen replied:
I don't see how this would cause a leak.  Yes, in some cases listeners can cause troubles, but I don't see how it is a problem here.  Can you be more specific?  Otherwise, I don't know what needs a fixin' here.

ottobackwards replied:
Sorry, battle scars.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/StellarExecutionNotifier.java:
```diff
@@ -0,0 +1,44 @@
+   * Add a listener that will be notified when a magic command is defined.
+   * @param listener The listener to notify.
+   */
+  void addSpecialListener(StellarExecutionListeners.SpecialDefinedListener listener);
```
Our `DefaultStellarShellExecutor` is a `StellarExecutionNotifier` as it is able to notify event listeners when variables, functions or specials are defined.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/StellarResult.java:
```diff
@@ -0,0 +1,185 @@
+/**
+ * The result of executing a Stellar expression within a StellarShellExecutor.
+ */
+public class StellarResult {
```
Instead of the Stellar executor just returning an Object like before (which doesn't tell me much about whether the operations was successful or not) I needed a more descriptive result like this.  

This is how the CLI and Zeppelin REPLs determine whether the Stellar expression was executed successfully or not.  They need to perform different actions based on success, failure or even if there is a termination request.

ottobackwards replied:
If this is Shell only can we name it StellarShellResult?

nickwallen replied:
That's what I had it originally.  Ha.  

It is already in the "shell" package, so I thought making the name simpler is just easier.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/specials/Comment.java:
```diff
@@ -0,0 +1,53 @@
+ *
+ *    # this is a comment
+ */
+public class Comment implements SpecialCommand {
```
Comments are not handled in the language itself.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/specials/AssignmentCommand.java:
```diff
@@ -0,0 +1,85 @@
+ *
+ *    x := 2 + 2
+ */
+public class AssignmentCommand implements SpecialCommand {
```
Assignment is not handled in the language (yet #687 ?)

nickwallen replied:
Right, it is not.  And that is one advantage of this design.  Once #687 gets merged, all we have to do is remove this class.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/specials/DocCommand.java:
```diff
@@ -0,0 +1,105 @@
+ *
+ * For example `?TO_STRING` will output the docs for the function `TO_STRING`
+ */
+public class DocCommand implements SpecialCommand {
```
Doc strings are not part of the core language.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/specials/QuitCommand.java:
```diff
@@ -0,0 +1,51 @@
+ *
+ *    quit
+ */
+public class QuitCommand implements SpecialCommand {
```
This is what allows a user to execute `quit` within the REPL.  Again, not part of core Stellar.

nickwallen commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/specials/SpecialCommand.java:
```diff
@@ -0,0 +1,60 @@
+ * This is typically an action performed on the execution
+ * environment, not something that could be executed within Stellar.
+ */
+public interface SpecialCommand {
```
The interface that I used for any logic that we had implemented in the REPL that is not part of the Stellar language.

nickwallen commented on metron-stellar/stellar-zeppelin/src/main/java/org/apache/metron/stellar/zeppelin/StellarInterpreter.java:
```diff
@@ -0,0 +1,158 @@
+/**
+ * A Zeppelin Interpreter for Stellar.
+ */
+public class StellarInterpreter extends Interpreter {
```
This is what allows us to run Stellar in a Notebook.

nickwallen commented on metron-stellar/stellar-zeppelin/src/main/java/org/apache/metron/stellar/zeppelin/StellarInterpreter.java:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  @Override
+  public List<InterpreterCompletion> completion(String buf, int cursor) {
```
This is where we connect our generic auto-completer that is used across both REPLs into the auto-complete extension point offered by Zeppelin.

ottobackwards commented:
This is quite a bit to go over.... can I trade you a similarly large review task?

merrimanr commented:
I think this is an excellent start.  So far I have only reviewed it from a user perspective and it's working well so far.  

I've spun it up in full dev (not sure that's even necessary) and installed this via the instructions in the README.  I was able to run most of the examples in the Stellar README.

I found that the functions available in Zeppelin are a subset of what's available in the Stellar shell.  The missing functions include IS_EMAIL, ENRICHMENT*, GEO*, STATS* and many others.  Is this expected?

Is there a way to pass in the zookeeper url in Zeppelin?

merrimanr commented:
I also noticed that functions returning a list only display the first item in both the shell and Zeppelin output.

For example, the expression `MAP([ 'foo', 'bar'], (x) -> TO_UPPER(x) )` returns `FOO` when I would expect it to return `['FOO', 'BAR']`.  

From what I can tell the problem is only in how the result is displayed.  For example, the expression `'BAR' in MAP([ 'foo', 'bar'], (x) -> TO_UPPER(x) )` returns `true` as expected.

nickwallen commented:
Thanks for digging into it @merrimanr .  You uncovered some good stuff.

> I found that the functions available in Zeppelin are a subset of what's available in the Stellar shell. The missing functions include IS_EMAIL, ENRICHMENT*, GEO*, STATS* and many others. Is this expected?

Yes, since we only added `stellar-common` to the interpreter, only the functions defined in that library are available.

I just updated the README to clarify this point.  I also added instructions for adding additional libraries to gain access to more Stellar functions.

nickwallen commented:
> @merrimanr: Is there a way to pass in the zookeeper url in Zeppelin?

No, I did not implement that in this PR.  As a next step I was going to do whatever needs done to get the management functions working in Zeppelin.  That would include adding a Zookeeper URL.

nickwallen commented:
> @merrimanr: From what I can tell the problem is only in how the result is displayed. For example, the expression 'BAR' in MAP([ 'foo', 'bar'], (x) -> TO_UPPER(x) ) returns true as expected.

BUG!  Good find.  The value getting returned is always the first.  I'll fix that.

nickwallen commented:
The problem is on the UI side of things; for both Zeppelin and the CLI.  When I get the result back from Stellar, I was using `ConversionUtils.convert(value, String.class)` to get me a result that I can display.  `ConversionUtils` just gives you the first item back if you ask it to convert a list to a String.  Oops.

nickwallen commented:
@merrimanr  Fixed that bug and was able to add a good amount of unit tests around it.  Also figured out a way to unit test the Aesh-driven StellarShell class.

merrimanr commented:
I spun this up again and all the issues I found have been resolved.  Great work.  +1

ottobackwards commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/DefaultStellarShellExecutor.java:
```diff
@@ -0,0 +1,413 @@
+            .with(SHELL_VARIABLES, () -> variables)
+            .with(STELLAR_CONFIG, () -> properties);
+
+    // load global configuration from zookeeper
```
the optional.ifPresent(()->) syntax is a nice style maybe

nickwallen replied:
I think it is simpler if I just use an 'if' statement.  I like things to be stupid simple.  'ifPresent' is nice when you have a 1 liner, but this takes quite a few lines of code, so I went with the 'if' conditional.  We do that in other places in the code base also.

ottobackwards commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/StellarResult.java:
```diff
@@ -0,0 +1,185 @@
+    this.status = status;
+    this.value = Optional.ofNullable(value);
+    this.exception = Optional.empty();
+    this.isValueNull = (value == null);
```
&& status == Status.SUCESS ?

nickwallen replied:
Ok, sure.  I'll fix that.

nickwallen replied:
There was no way for this to cause a problem because the constructors are private.  But I added unit tests and your suggested check just to be sure.

ottobackwards commented on metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/StellarShellExecutor.java:
```diff
@@ -0,0 +1,79 @@
+
+  /**
+   * The current state of the Stellar execution environment.
+   */
```
Are variables the only state we can have?  I'm going to have to change this when I write the debugger.

nickwallen replied:
Yes.

nickwallen commented:
> Overall: I would like to see this wait until we make some effort to clear outstanding stellar PR's from the queue. Making PR authors whose PR's don't get reviewed continually update to track changes as we focus on the 'new' things doesn't seem right. I am totally not biased because I'm the guy that has to do that either.

I think you need to bring this up as  a discuss thread.  

@ottobackwards Are you -1'ing this PR then?

ottobackwards commented:
I would not, and am not -1 a pr for non-technical reasons.  I'm not ready to +1 this, since I have outstanding comments and actually have to try it out beyond scanning the diffs.

ottobackwards commented:
When I say I would like a readme or outline document, here is where I'm coming from:  As someone who does do stellar and stellar shell work, I would like to understand the high level view of what you have put in place, and have that documented beyond the context of the pr, which is "I had to do this and touch this".   This would  help keep myself and others from regressing one or the other.

Also, do we now have to validate in both environments when we make changes to the shell?  that should also be spelt out.

nickwallen commented:
> When I say I would like a readme or outline document, here is where I'm coming from: As someone who does do stellar and stellar shell work, I would like to understand the high level view of what you have put in place, and have that documented beyond the context of the pr, which is "I had to do this and touch this". This would help keep myself and others from regressing one or the other.

You want the README to be like a design document?  Where should this README live?

nickwallen commented:
> Also, do we now have to validate in both environments when we make changes to the shell? that should also be spelt out.

No you do not have to validate in both.  The point of all this refactoring was so that we don't have duplicate logic between the CLI REPL and Zeppelin.

(1) There is the `shell.StellarShell` that has all the Aesh code to drive the CLI shell.
(2) There is the `zeppelin.StellarInterpreter` that has all the logic to interact with Zeppelin.
(3) Finally there is the `shell.StellarShellExecutor` that is used by both (1) and (2) to handle Stellar and our Stellar "extensions" like assignment, comments, etc.

All of this now has a plethora of unit tests; tests that we didn't have before.  So I think this puts us in a better position to avoid future regressions.

ottobackwards commented:
I don't thin a full design doc is what we need, just an overview.  It should be in the top level of common.shell. 
..........
Stellar can be hosted in two shell like environments, zeppelin and the cli.  This packaged is factored to share the implementation details of the shell, while allowing specific integration mechanisms.

The common stuff can be found HERE
This includes:

- Commands, which are things that do things like A,B,C
- Autocomplete, which is implemented commonly HERE, and brought into the specific implementation by FOO
- The executors
etc

AESH specific notes:


Zeppelin specific notes:


---------------

nickwallen commented:
Ok will do @ottobackwards

nickwallen commented:
And done.  Please take a look when you get a chance.

ottobackwards commented:
Is there a 'supported' or tested version of zeppelin?
I have 0.7.3 installed locally.

I am getting exceptions starting the stellar interpreter

```
org.apache.zeppelin.interpreter.InterpreterException: org.apache.thrift.TApplicationException: Internal error processing getFormType
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.getFormType(RemoteInterpreter.java:450)
	at org.apache.zeppelin.interpreter.LazyOpenInterpreter.getFormType(LazyOpenInterpreter.java:111)
	at org.apache.zeppelin.notebook.Paragraph.jobRun(Paragraph.java:387)
	at org.apache.zeppelin.scheduler.Job.run(Job.java:175)
	at org.apache.zeppelin.scheduler.RemoteScheduler$JobRunner.run(RemoteScheduler.java:329)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.thrift.TApplicationException: Internal error processing getFormType
	at org.apache.thrift.TApplicationException.read(TApplicationException.java:111)
	at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:71)
	at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Client.recv_getFormType(RemoteInterpreterService.java:337)
	at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Client.getFormType(RemoteInterpreterService.java:323)
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.getFormType(RemoteInterpreter.java:446)
	... 11 more
```

Also:

In the zeppelin site file, there is already a zeppelin.interpreters property, which in a comma separated list of class names.  Is it right to create a new property vs. adding to the existing?

ottobackwards commented:
Another question:
Even through it is consistent with the other interpreters, I am not sure that we should not have Apache Stellar, or Apache Metron Stellar and not just stellar as the name.  It does not seem right to not have Apache there.

nickwallen commented:
> Is there a 'supported' or tested version of zeppelin? I have 0.7.3 installed locally.

0.7.3 is what the unit tests and what my manual testing have used.  

I added this to the README.

nickwallen commented:
> @ottobackwards : In the zeppelin site file, there is already a zeppelin.interpreters property, which in a comma separated list of class names. Is it right to create a new property vs. adding to the existing?

Both work, but adding it to the existing is probably the better approach.  I updated the README to reflect this.

nickwallen commented:
> @ottobackwards : I am getting exceptions starting the stellar interpreter

I have never seen this.  I need more context unfortunately.

1. What do you mean by "starting the stellar interpreter"?  

1. Is the exception coming from the Zeppelin logs or are you seeing this in a notebook?

1. When do you see the exception?  After starting Zeppelin or after running Stellar in a notebook?  Or sometime else?

1. Are you successfully running Stellar in a notebook at all?

1. I am assuming you followed the instructions in the README exactly?  Is there anything in there that might be different from what you did?

ottobackwards commented:
1.  When I type something in the notebook and hit run.  At least I think that is when I get the error.  It may happen when selecting the interpreter for the notebook.

2. Yes

3. Running TO_UPPER('foo')

4. no, I get this for every statement I run.
Also, I don't get %stellar for the prompt like your screenshots

5.  
 - I set the interpreter through the dropdown not by typing %stellar
- I didn't have to 'create' the stellar interpreter after starting, there was already an entry, I think from the site config

nickwallen commented:
> I didn't have to 'create' the stellar interpreter after starting, there was already an entry, I think from the site config

That does not (should not) happen.  It should only be there when after you have created it in the UI.

Are you sure you haven't tried this a few times and either...
(1) There are left-over files in $ZEPPELIN_HOME between attempts
(2) Something is stored in browser cache, or even that... 
(3) The Zeppelin instance was launched (and remains running) from one of your first attempts?

When I was playing with this a few weeks ago and was seeing weird behavior, I found that (3) was a problem for me.  I either left the Zeppelin process running accidentally or the stop script in Zeppelin does not work all the time.  I had to `ps -ef | grep Zep` and forcibly kill processes to make sure that the process was running with the configuration I expected.

ottobackwards commented:
I think creating the /interpreters/stellar/x.json creates the entry?

I just tried again, ( new mvn install , new site file ) and I checked the interpreter log:

INFO [2018-01-09 16:14:22,873] ({Thread-0} RemoteInterpreterServer.java[run]:97) - Starting remote interpreter server on port 63321
   2 ERROR [2018-01-09 16:14:23,545] ({pool-1-thread-3} RemoteInterpreterServer.java[createInterpreter]:204) - java.lang.ClassNotFoundException: org.apache.metron.stellar.zepplin.StellarInterpreter
   3 java.lang.ClassNotFoundException: org.apache.metron.stellar.zepplin.StellarInterpreter
   4 >   at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
   5 >   at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
   6 >   at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
   7 >   at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
   8 >   at java.lang.Class.forName0(Native Method)
   9 >   at java.lang.Class.forName(Class.java:264)
  10 >   at org.apache.zeppelin.interpreter.remote.RemoteInterpreterServer.createInterpreter(RemoteInterpreterServer.java:178)
  11 >   at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Processor$createInterpreter.getResult(RemoteInterpreterService.java:1662)
  12 >   at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Processor$createInterpreter.getResult(RemoteInterpreterService.java:1647)
  13 >   at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)
  14 >   at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
  15 >   at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:285)
  16 >   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  17 >   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  18 >   at java.lang.Thread.run(Thread.java:745)

ottobackwards commented:
Can you run the check style against the interpreter class?  
Also, do we want to use `@Override`?

nickwallen commented:
> Also, do we want to use @Override?

Done.

ottobackwards commented:
@nickwallen I'm not sure what to do to get this going.  The jar is in maven.  I tried to path directly to it as well.  I can attach my site and interpreter json files, and screenshot my settings....

ottobackwards commented:
I'm going to try a clean zeppelin install I guess.

ottobackwards commented:
ok, starting over

1. First difference, is on first start ( new install ) the stellar interpreter / group is already created in zeppelin when I go to configure the stellar interpreter

<img width="1393" alt="screen shot 2018-01-10 at 09 32 02" src="https://user-images.githubusercontent.com/1111551/34777713-30d04262-f5e9-11e7-9345-f8588b404301.png">

So I update and restart with new settings.

Interpreter is green.

2. Run the first two tasks with the following results:
<img width="1387" alt="screen shot 2018-01-10 at 09 39 20" src="https://user-images.githubusercontent.com/1111551/34778077-34c44304-f5ea-11e7-8921-f1d32102bb1e.png">

3. I was able to run other stellar statements fine, so it is better

nickwallen commented:
Ok, so it looks like it is working for you.

Zeppelin interprets the first line as the interpreter spec (or whatever they call it).  That's why `%functions` failed because it thinks you want to use an interpreter called 'functions'.  If you want to use a magic function like `%functions` either put that on a second line or make sure to specify `%stellar` explicitly like so...

```
%stellar

%functions
```

ottobackwards commented:
Is the plan to deploy the jar file in the interpreter directory like the other interpreters?

nickwallen commented:
> Is the plan to deploy the jar file in the interpreter directory like the other interpreters?

I don't know what that is going to look like.  We need to do more research to see how the setup can be automated by the Mpack.  I don't want the user to have to do all this manual setup.

ottobackwards commented:
Do we have jiras for the next steps on this yet?  I would be good if we recored our intent, and who knows, maybe someone will want to work on it.

ottobackwards commented:
This is really good work @nickwallen.  I am +1 on this, but would like to see jiras on the next steps entered if possible as a side note.
